packages/marshal/src/encodePassable.js

```diff
@@ -0,0 +1,419 @@
+// numeric sort order of the corresponding numbers, the characters of the unary
+// prefix are different for negative values (type "n" followed by any number of
+// "#"s [which sort before decimal digits]) vs. positive and zero values (type
+// "p" followed by any number of "~"s [which sort after decimal digits]) and
+// each decimal digit of the encoding for a negative value is replaced with its
+// ten's complement (so that negative values of the same scale sort by
+// *descending* absolute value).
+const encodeBigInt = n => {
+  const abs = n < 0n ? -n : n;
+  const nDigits = abs.toString().length;
+  const lDigits = nDigits.toString().length;
+  if (n < 0n) {
+    return `n${
+      // A "#" for each digit beyond the first
+      // in the decimal *count* of decimal digits.
+      '#'.repeat(lDigits - 1)
+    }${
+      // The ten's complement of the count of digits.
+      (10 ** lDigits - nDigits).toString().padStart(lDigits, '0')
+    }:${
+      // The ten's complement of the digits.
+      (10n ** BigInt(nDigits) + n).toString().padStart(nDigits, '0')
+    }`;
+  } else {
+    return `p${
+      // A "~" for each digit beyond the first
+      // in the decimal *count* of decimal digits.
+      '~'.repeat(lDigits - 1)
+    }${
+      // The count of digits.
+      nDigits
+    }:${
+      // The digits.
+      n
+    }`;
+  }
+};
+
+const decodeBigInt = encoded => {
+  const typePrefix = encoded[0];
+  let rem = encoded.slice(1);
+  assert(
+    typePrefix === 'p' || typePrefix === 'n',
+    X`Encoded bigint expected: ${encoded}`,
+  );
+
+  const lDigits = rem.search(/[0-9]/) + 1;
+  assert(lDigits >= 1, X`Digit count expected: ${encoded}`);
+  rem = rem.slice(lDigits - 1);
+
+  assert(rem.length >= lDigits, X`Complete digit count expected: ${encoded}`);
+  const snDigits = rem.slice(0, lDigits);
+  rem = rem.slice(lDigits);
+
+  assert(
+    /^[0-9]+$/.test(snDigits),
+    X`Decimal digit count expected: ${encoded}`,
+  );
+  let nDigits = parseInt(snDigits, 10);
+  if (typePrefix === 'n') {
+    // TODO Assert to reject forbidden encodings
+    // like "n0:" and "n00:…" and "n91:…" through "n99:…"?
+    nDigits = 10 ** lDigits - nDigits;
+  }
+
+  assert(rem.startsWith(':'), X`Separator expected: ${encoded}`);
+  rem = rem.slice(1);
+
+  assert(
+    rem.length === nDigits,
+    X`Fixed-length digit sequence expected: ${encoded}`,
+  );
+  let n = BigInt(rem);
+  if (typePrefix === 'n') {
+    // TODO Assert to reject forbidden encodings
+    // like "n9:0" and "n8:00" and "n8:91" through "n8:99"?
+    n = -(10n ** BigInt(nDigits) - n);
+  }
+
+  return n;
+};
+
+// `'\u0000'` is the terminator after elements.
+// `'\u0001'` is the backslash-like escape character, for
+// escaping both of these characters.
+
+const encodeArray = (array, encodePassable) => {
+  const chars = ['['];
+  for (const element of array) {
+    const enc = encodePassable(element);
+    for (const c of enc) {
```

Comment:
That's a lot of iterating strings, one character at a time, repeated for every depth of the input object graph, and double that for records. (`{ foo: {bar: 'baz'} }` is going to see `foo` walked twice and `bar` walked four times. Let's be sure to benchmark this: I suspect this string walk will dominate the encoding time.

The alternative might be a length-prefix on each component string, which also avoids the exponential escape-the-escape-character penalty. Even if the caller (and the other layers of protocol they might be using) manages to avoid calling `marshal.serialize` twice on the same data, the fact that `encodeArray` needs to escape the output of `encodePassable` means that even e.g. `[ [ 1, 2 ] ]` is going to invoke the `\u0001` escape, and `[ [...Array(101)] ]` is going to have 100 `\u0001\u0000` sequences, and `[ [ [...Array(101)] ] ]` will have 200, and `[ [ [ [...Array(101)] ] ] ]` will have 400, etc. Every record in the nesting will cost double.

Reply:
Attn @gibson042

Reply:
I believe this could be fixed rather conveniently by moving the escaping logic from _array_ encoding/decoding to _string_ encoding/decoding and refactoring array decoding to handle depth. For example, assuming the array sigil is updated from "[" to "~" for control-character adjacency:
* Define string escapes like `const escapes = { __proto__: null, "\0": "\u{0001}0", "\x01": "\u{0001}1", "~": "\u{007F}0", "\x7F": "\u{007F}1" }` (or to squeeze out every bit of performance, replace that object with an array like `escapes = Array(128).fill().map((_, codePoint) => escapes[String.fromCodePoint(codePoint)])` or its sparse equivalent).
* Encode strings like ``` `s${passable.replaceAll(/[\0\x01~\x7F]/g, c => escapes[c])}` ```.
* Decode arrays like
  ```js
  const decodeArray = (encoded, decodePassable) => {
    encoded.startsWith('~') || assert.fail(X`invalid array encoding: ${encoded}`);
    const tail = encoded.slice(1);
    let elements = [];
    let depth = 0;
    let nextIndex = 0;
    let currentElementIndex = 0;
    for (let { 0: c, index } of tail.matchAll(/[\0~]/g)) {
      if (c === '~') {
        // This is the start of a nested array.
        depth += 1;
      } else {
        // This is an element terminator.
        if (index === nextIndex) {
          // A terminator after ~ or another terminator ends the deepest array.
          depth -= 1;
          (depth >= 0) ||
            assert.fail(X`unexpected element terminator: ${encoded.slice(0, index + 2)}`);
        }
        if (depth === 0) {
          // We have a complete top-level element.
          elements.push(decodePassable(tail.slice(currentElementIndex, index)));
          currentElementIndex = index + 1;
        }
      }
      // Advance the index.
      nextIndex = index + 1;
    }
    (depth === 0) || assert.fail(X`unterminated encoded array: ${encoded}`);
    (nextIndex === tail.length) ||
      assert.fail(X`unterminated encoded array element: ${tail.slice(currentElementIndex)}`);
    return harden(elements);
  };
  ```
  (and note that a further optimization is possible by decoding nested arrays into a cache rather than discarding their contents after syntax validation and thereby forcing repeat work for each increased level)

Sample encodings:
* `[]`: "~"
* `[[]]`: "~~\u0000"
* `'foo'`: "sfoo"
* `'\0,\x01,~,\x7F'`: "s\u00010,\u00011,\u007F0,\u007F1"
* `['\0']`: "~s\u00010\u0000"
* `[['\0']]`: "~~s\u00010\u0000\u0000"
* `['~']`: "~s\u007F0\u0000"
* `[['~']]`: "~~s\u007F0\u0000\u0000"
* `{ foo: "bar" }`: "(~~sfoo\u0000\u0000~sbar\u0000\u0000"

Reply:
And we can also reduce the size of JSON representations by abandoning characters between U+0000 and U+001F (inclusive), which require "\u…" escape sequences. This would look something like changing the array element terminator from U+0000 to "&" (updating string escapes to include "&" → "%1" and "%" → "%0") and—because there is a requirement that the terminator sorts before anything else—the error sigil from "!" to "*" (assuming we don't care about the relative sort order of types w.r.t. each other).
* `[]`: "~"
* `[[]]`: "~~&"
* `'foo'`: "sfoo"
* `'\0,\x01,%,&,~,\x7F'`: "s\u0000,\u0001,%0,%1,␡0,␡1"
* `['\0']`: "~s\u0000&"
* `[['\0']]`: "~~s\u0000&&"
* `['~']`: "~s␡0&"
* `[['~']]`: "~~s␡0&&"
* `{ foo: "bar" }`: "(~~sfoo&&~sbar&&"

Reply:
@gibson042 I decided not to make these format changes as part of this PR. Instead I filed https://github.com/endojs/endo/issues/1349 and assigned both of us to remember to do it later.

Resolving.